### PR TITLE
Improvements for site pin and bel pin connections in Wire class

### PR DESCRIPTION
**Pull request by trharoldsen**
Improved support for connecting to BelPin and SitePins in the Wire class.  The class now has methods that will return a single SitePin/BelPin that a wire connects to instead of a collection of Terminal/Pin connections.  The old methods still exist but have been marked as deprecated along with the related Connection subclasses and Connection class methods relating to determining the type of connection.  All connections now are wire connections.  

The SiteWire now contains a SiteType specifying the type of the site the wire was created with.  Thus if a site changes, the connections will no longer return empty.  This will allow treating the routing and device as immutable objects if desired and not needing to worry about whether the type of the site has changed.  To instrument this, a new method has been added to the Wire class getAllConnectedPins() that will get the SitePin object for each representation of a Site that the Wire connects to.

In addition, some outdated and unused methods from the Wire classes and Connection classes have been deleted.  Lastly, all usages inside of RapidSmith of these deprecated features have been updated.

Added the following methods to wire and its subclasses:
  getSource()
  getTerminal()
  getConnectedPin()
  getReverseConnectedPin()

and deprecated:
  getSources()
  getTerminals()
  getPinConnections()
  getReversePinConnections()

Deleted getAllWireConnections and getAllReverseWireConnections from Wire class
Added getAllConnectedPins and getAllReverseConnectedPins to Wire class

Deprecated TileToSite, SiteToTile and Terminal Connections classes
Deprecated the following methods on Connections class:
  isPinConnection()
  isTerminal()
  getSitePin()
  getBelPin()
  getWireConnections()
  getPinConnections()
  getTerminals()
  getAllConnections()

### CODE REVIEW

**Comment by trharoldsen** [comment]
Note to @mcrossen, some of the methods in RouteTree have been updated.  Their signatures should be the same though.

File: src/main/java/edu/byu/ece/rapidSmith/device/SitePin.java [hunk 1]
```diff
@@ -65,6 +65,15 @@ public SiteType getSiteType() {
 	}
 
 	/**
+	 * Provides access to the internal template backing this object.  This method
+	 * is not needed for normal use.
+	 * @return the template backing this object
+	 */
+	public SitePinTemplate getTemplate() {
```
Comment (ttown523) [inline src/main/java/edu/byu/ece/rapidSmith/device/SitePin.java]: If its not needed for normal use, can we make it package private?

File: src/main/java/edu/byu/ece/rapidSmith/device/Wire.java [hunk 2]
```diff
@@ -38,45 +37,73 @@
...
-
-	/**
 	 * Return connection linking this wire to other wires in the same hierarchy.
 	 */
 	Collection<Connection> getWireConnections();
 
 	/**
 	 * Returns connection linking this wire to another wire in a different
 	 * hierarchical level through a pin.
+	 * @deprecated
```
Comment (ttown523) [inline src/main/java/edu/byu/ece/rapidSmith/device/Wire.java]: these should be moved outside the comment block

File: src/main/java/edu/byu/ece/rapidSmith/device/Wire.java [hunk 3]
```diff
@@ -38,45 +37,73 @@
...
 	/**
 	 * Returns connection linking this wire to its drivers in the same hierarchy.
 	 */
 	Collection<Connection> getReverseWireConnections();
 
 	/**
+	 * Returns the connected site pins for each possible type of the connected site.
+	 * @return all connected sites pins of this wire
+	 */
+	Collection<SitePin> getAllReverseSitePins();
```
Comment (ttown523) [inline src/main/java/edu/byu/ece/rapidSmith/device/Wire.java]: should we deprecate this too?

File: src/main/java/edu/byu/ece/rapidSmith/device/Wire.java [hunk 4]
```diff
@@ -38,45 +37,73 @@
...
 
 	/**
+	 * Return connection linking this wire to its drivers in the different
+	 * levels of hierarchy.
+	 */
+	SitePin getReverseConnectedPin();
+
+	/**
 	 * Returns the sources (BelPins) which drive this wire.
+	 * @deprecated
```
Comment (ttown523) [inline src/main/java/edu/byu/ece/rapidSmith/device/Wire.java]: move this outside comment block

**Pull request merged**